### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized pool/forest predictor model with float arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#366]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(29801280317071).epochs(1).f(6, 9).log(all).stop(0.8).run()` on `backend:nv0` exceeds the 60-second threshold. The model (`recipe.model().pool(2).exp().norm(batch).iq(4).nl.forest(2).huber().norm(batch).qi(5).0.loss(mae)`) reaches `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2852-2896) before any epoch arithmetic. The dataset (`data/temporal/chronological_splits`) has ~12000 rows and 1 feature; the workload (pool size 2, forest with 2 trees, no boosting iterations) justifies sub-second throughput on nv0, so the measured 61s stable hang is the uncached native artifact compilation, not arithmetic or memory traffic. No existing nv0 stall issue covers this quantized pool plus predictor (forest) primitive pair (291 = pool/GRU/scan, 282 = predictor-only); the loader and `.f(6,9)` arithmetic execute at or after the failure point.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `1bfffb2dd734147c`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized pool/forest predictor model with float arithmetic",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.001).seed(29801280317071).epochs(1).f(6, 9).log(all).stop(0.8).run()` on `backend:nv0` exceeds the 60-second threshold. The model (`recipe.model().pool(2).exp().norm(batch).iq(4).nl.forest(2).huber().norm(batch).qi(5).0.loss(mae)`) reaches `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2852-2896) before any epoch arithmetic. The dataset (`data/temporal/chronological_splits`) has ~12000 rows and 1 feature; the workload (pool size 2, forest with 2 trees, no boosting iterations) justifies sub-second throughput on nv0, so the measured 61s stable hang is the uncached native artifact compilation, not arithmetic or memory traffic. No existing nv0 stall issue covers this quantized pool plus predictor (forest) primitive pair (291 = pool/GRU/scan, 282 = predictor-only); the loader and `.f(6,9)` arithmetic execute at or after the failure point.",
  "rationale": "Inspected `recipe.rs` source confirms the stall reaches `compile_native_artifact` before epoch work begins (`recipe.rs` 2868-2903, invoked from `compile_model` at 2870). The reproduction lowers `.pool(2)` to `Primitive::Pool` (`lower_pool`, 5115-5308) and `.forest(2)` to `Primitive::Predictor` (`lower_estimator`, 5464-5489, `fit_forest` 7207-7230), so the emitted native IR contains both pool and predictor primitives, a pairing unrepresented among open nv0 stall issues (271 residual, 282 predictor primitives, 291 pool/scan, 308 pool/attention, 315 perc/MoE, etc.). Workload derivation: `data/temporal/chronological_splits/train.csv` has ~12000 rows with 1 input feature; model operations are limited to a size-2 pool and a 2-tree predictor with no additional estimator expansion; memory traffic is under tens of MB. Comparable nv0 epochs are measured in milliseconds, so 61s exceeds justifiable throughput by roughly two orders of magnitude with the dominant gap being the NVIDIA native compiler hang. Because the earliest failing operation (`compile_native_artifact`) is the same terminal check but reached through a materially different primitive pair (`Pool` + `Predictor` vs existing `Pool`+`Scan` or predictor-only), this input representation establishes a different earliest cause rather than additional evidence for an existing issue."
}
```

## Deterministic failure packet

kind=performance
id=1bfffb2dd734147c
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:11189 next:11190 composition:29801280317071
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 61 seconds
measurement=elapsed_seconds:61 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-132-11189/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.pool(2).exp().norm(batch).iq(4).nl
		.forest(2).huber().norm(batch).qi(5).0
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(29801280317071)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Failed on Thursday, August 27, 2026 2:35 PM PDT

Commit: 8f386573e4b6979d7e4d03204602144888a2ebe6

Command:
```text
recipe test.rs
```

Reproduction script:
```rust
use recipe::*;
use std::time::{Duration, Instant};

fn main() {
	let bundle = format!("/tmp/recipe-issue-366-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.pool(2).exp().norm(batch).iq(4).nl
		.forest(2).huber().norm(batch).qi(5).0
		.loss(mae);
	let started = Instant::now();
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(29801280317071)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	let elapsed = started.elapsed();
	let final_loss = report.final_loss();
	println!("final_loss_finite={}", final_loss.is_finite());
	println!("elapsed_seconds={:.3}", elapsed.as_secs_f64());
	println!("within_60_seconds={}", elapsed < Duration::from_secs(60));
	assert!(final_loss.is_finite());
	assert!(elapsed < Duration::from_secs(60));
}
```

Full output:
```text
recipe: unknown command "test.rs"; expected `probe`

Recipe command line

Usage:
  recipe probe [OPTIONS]

The zero-argument probe uses the embedded theoretical seed contract, discovers
the current bare-metal machine, benchmarks every discovered device and link,
and writes an identity-keyed measured profile under the user's private cache.

Options:
  --contract PATH       Use another theoretical probe seed contract
  --profile PATH        Write/load this absolute measured-profile path
  --cuda-driver PATH    Exact CUDA Driver library candidate (repeatable)
  --hsa-runtime PATH    Exact ROCr/HSA runtime candidate (repeatable)
  --llvm-opt PATH       Exact LLVM IR verifier
  --llvm-llc PATH       Exact LLVM code generator
  --lld PATH            Exact ELF linker
  --ptxas PATH          Exact NVIDIA PTX assembler
  -h, --help             Show this help

```

Verdict: FAIL
Reason: The required public `recipe test.rs` path is absent because the `recipe` executable accepts only `probe`. The reproduction never reaches model compilation or the one-epoch nv0 run, so finite loss, completion within 60 seconds, and throughput proportionate to the stated workload remain unproven.

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #456.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #456: Verify current NVIDIA pool forest path</summary>

Closes #366

The exact current public NVIDIA reproduction completes within the reported 60-second boundary. No issue-specific production change is supported by the current behavior.

Exact command:

```sh
/usr/bin/time -v env RECIPE_DEBUG=1 ./target/release/recipe --device benji:nv0 /home/nate/.cache/recipe/issues/366/reproduction.rs
```

The directory contains 18,000 logical rows, not approximately 12,000: 12,000 train rows, 3,000 validation rows, and 3,000 test rows. The single raw timestamp column is encoded as `Text(20)`. The parser correctly ignores 490,537 trailing blank lines in the test file.

Cold-cache result:

- Wall time: 19.55 seconds
- Epoch time: 772.399 ms
- MAE: 24.5082
- Maximum resident memory: 111,576 KB
- Exit status: 0

The cold debug log records three native cache misses. Each completes Clang bitcode compilation, `llc` PTX generation, native loading, and its required forward or epoch path. The final quantized pool and forest graph completes its epoch and metric calculation.

Warm-cache result:

- Wall time: 16.02 seconds
- Epoch time: 764.221 ms
- MAE: 24.5082
- Maximum resident memory: 111,388 KB
- Exit status: 0

The cold and warm bundles are both 369,710 bytes and have the identical SHA-256 digest `e8678440664eb263927c62fd990f985ebbc90be43a4b3032a6874a170ca517c6`.

The issue's only prior comment ran a probe-only installed binary with `recipe test.rs`; that command was rejected before loading data, compiling the model, or running an epoch. Current `minimal` instead uses the merged Clang bitcode and `llc` PTX path exercised above.

Validation also passed `cargo fmt --check`, `git diff --check`, `cargo check`, and `cargo build --release --bin recipe`. This verification commit intentionally changes no tracked files.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
